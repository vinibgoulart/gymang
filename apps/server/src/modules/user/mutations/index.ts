import UserAdd from './UserAddMutation';
import UserLogin from './UserLoginMutation';
import UserLogout from './UserLogoutMutation';

const UserMutations = {
  UserAdd,
  UserLogin,
  UserLogout,
};

export default UserMutations;
