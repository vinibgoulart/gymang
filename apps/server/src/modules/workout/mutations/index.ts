import WorkoutAdd from './WorkoutAddMutation';

const WorkoutMutations = {
  WorkoutAdd,
};

export default WorkoutMutations;
