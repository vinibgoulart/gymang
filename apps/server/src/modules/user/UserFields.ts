import { GraphQLNonNull } from 'graphql';
import UserType, { UserConnection } from './UserType';
import { UserLoader } from '@gymang/user';
import {
  edgeField,
  connectionArgs,
  NullConnection,
  withFilter,
} from '@gymang/graphql';
import { isLoggedIn } from '@gymang/core';

export const userTypeField = (key = 'user', bypassViewerCanSee = false) => ({
  [key]: {
    type: UserType,
    resolve: async (obj, args, context) => {
      return UserLoader.load(context, obj[key], bypassViewerCanSee);
    },
  },
});

export const userEdgeField = () =>
  edgeField({
    connection: UserConnection,
    load: UserLoader.load,
    edgeName: 'userEdge',
    name: 'User',
  });

export const userConnectionField = (customResolver = null) => ({
  users: {
    type: new GraphQLNonNull(UserConnection.connectionType),
    args: {
      ...connectionArgs,
      // filters: {
      //   type: UserFilterInputType,
      // },
    },
    resolve: (obj, args, context) => {
      if (customResolver) {
        return customResolver(obj, args, context);
      }

      if (!isLoggedIn(context)) {
        return NullConnection;
      }

      const argsWithCompany = withFilter(args, {
        merchant: context?.merchant?._id,
      });

      return UserLoader.loadAll(context, argsWithCompany);
    },
  },
});

export const meField = () => ({
  me: {
    type: UserType,
    resolve: (_root, _args, context) => {
      return UserLoader.load(context, context.user?._id);
    },
  },
});
