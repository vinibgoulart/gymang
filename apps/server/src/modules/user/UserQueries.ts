import { meField, userConnectionField } from './UserFields';

const UserQueries = {
  ...meField(),
  ...userConnectionField(),
};

export default UserQueries;
