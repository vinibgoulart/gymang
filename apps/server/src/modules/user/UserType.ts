import { GraphQLContext } from '@gymang/core';
import {
  connectionDefinitions,
  nodeInterface,
  registerTypeLoader,
} from '@gymang/graphql';
import { IUser, UserLoader } from '@gymang/user';
import { GraphQLObjectType, GraphQLString } from 'graphql';
import { globalIdField } from 'graphql-relay';

const UserType = new GraphQLObjectType<IUser, GraphQLContext>({
  name: 'User',
  description: 'Represents a User',
  fields: () => ({
    id: globalIdField('User'),
    firstName: {
      type: GraphQLString,
      resolve: (user) => user.firstName,
    },
    email: {
      type: GraphQLString,
      resolve: (user) => user.email,
    },
  }),
  interfaces: () => [nodeInterface],
});

export const UserConnection = connectionDefinitions({
  name: 'User',
  nodeType: UserType,
});

registerTypeLoader(UserType, UserLoader.load);

export default UserType;
