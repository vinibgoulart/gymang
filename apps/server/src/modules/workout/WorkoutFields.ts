import type { GraphQLContext } from '@gymang/core';
import { isLoggedIn } from '@gymang/core';
import {
  edgeField,
  connectionArgs,
  NullConnection,
  withFilter,
} from '@gymang/graphql';
import { WorkoutFilterInputType, WorkoutLoader } from '@gymang/workout';
import { GraphQLNonNull } from 'graphql';
import type { ConnectionArguments } from 'graphql-relay';

import WorkoutType, { WorkoutConnection } from './WorkoutType';

export const workoutTypeField = <T>(
  key = 'workout',
  bypassViewerCanSee = false,
) => ({
  [key]: {
    type: WorkoutType,
    resolve: async (obj: T, _, context: GraphQLContext) => {
      return WorkoutLoader.load(context, obj[key], bypassViewerCanSee);
    },
  },
});

export const workoutEdgeField = () =>
  edgeField({
    connection: WorkoutConnection,
    load: WorkoutLoader.load,
    edgeName: 'workoutEdge',
    name: 'Workout',
  });

type WorkoutConnectionArgs = {
  filters: WorkoutFilterInputType;
} & ConnectionArguments;

export const workoutConnectionField = <T>(customResolver = null) => ({
  workouts: {
    type: new GraphQLNonNull(WorkoutConnection.connectionType),
    args: {
      ...connectionArgs,
      filters: {
        type: WorkoutFilterInputType,
      },
    },
    resolve: (obj: T, args: WorkoutConnectionArgs, context: GraphQLContext) => {
      if (customResolver) {
        return customResolver(obj, args, context);
      }

      if (!isLoggedIn(context)) {
        return NullConnection;
      }

      const argsWithFilter = withFilter(args, {
        isPublic: true,
      });

      return WorkoutLoader.loadAll(context, argsWithFilter);
    },
  },
});

type MeWorkoutConnectionArgs = ConnectionArguments;

export const meWorkoutConnectionField = <T = unknown>(
  customResolver = null,
) => ({
  meWorkouts: {
    type: new GraphQLNonNull(WorkoutConnection.connectionType),
    args: {
      ...connectionArgs,
      // filters: {
      //   type: WorkoutFilterInputType,
      // },
    },
    resolve: (
      obj: T,
      args: MeWorkoutConnectionArgs,
      context: GraphQLContext,
    ) => {
      if (customResolver) {
        return customResolver(obj, args, context);
      }

      if (!isLoggedIn(context)) {
        return NullConnection;
      }

      const argsWithUser = withFilter(args, {
        user: context.user._id,
      });

      return WorkoutLoader.loadAll(context, argsWithUser);
    },
  },
});
