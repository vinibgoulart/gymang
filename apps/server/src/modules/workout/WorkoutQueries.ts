import { myWorkoutConnectionField } from './WorkoutFields';

const WorkoutQueries = {
  ...myWorkoutConnectionField(),
};

export default WorkoutQueries;
