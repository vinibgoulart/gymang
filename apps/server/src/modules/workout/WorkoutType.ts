import type { GraphQLContext } from '@gymang/core';
import {
  connectionDefinitions,
  nodeInterface,
  registerTypeLoader,
} from '@gymang/graphql';
import type { IWorkout } from '@gymang/workout';
import { WorkoutLoader } from '@gymang/workout';
import { GraphQLNonNull, GraphQLObjectType, GraphQLString } from 'graphql';
import { globalIdField } from 'graphql-relay';

import { userTypeField } from '../user/UserFields';

const WorkoutType = new GraphQLObjectType<IWorkout, GraphQLContext>({
  name: 'Workout',
  description: 'Represents a Workout',
  fields: () => ({
    id: globalIdField('Workout'),
    name: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: (workout) => workout.name,
    },
    description: {
      type: GraphQLString,
      resolve: (workout) => workout.description,
    },
    ...userTypeField('createdBy'),
    ...userTypeField(),
  }),
  interfaces: () => [nodeInterface],
});

export const WorkoutConnection = connectionDefinitions({
  name: 'Workout',
  nodeType: WorkoutType,
});

registerTypeLoader(WorkoutType, WorkoutLoader.load);

export default WorkoutType;
