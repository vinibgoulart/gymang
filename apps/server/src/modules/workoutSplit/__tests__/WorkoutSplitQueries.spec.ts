import { GRAPHQL_TYPE } from '@gymang/core';
import { WORKOUT_SPLIT_MODALITY } from '@gymang/enums';
import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  sanitizeTestObject,
} from '@gymang/testutils';
import { handleCreateUser } from '@gymang/user';
import { handleCreateWorkout } from '@gymang/workout';
import { handleCreateWorkoutSplit } from '@gymang/workout-split';
import { graphql } from 'graphql';
import { toGlobalId } from 'graphql-relay';

import { getContext } from '../../../getContext';
import { schema as schemaAdmin } from '../../../graphql/schema/schema';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

const query = `
  query WorkoutSplitQueriesSpecQuery {
    meWorkoutSplits(first: 10) {
      edges {
        node {
          id
          name
          user {
            id
          }
          workout {
            id
            name
          }
          modality
        }
      }
    }
  }
`;

it('should get a list of workouts splits', async () => {
  const user = await handleCreateUser();
  const workout = await handleCreateWorkout({
    user,
    createdBy: user,
  });

  await handleCreateWorkoutSplit({
    name: 'A',
    workout,
    user,
    modality: WORKOUT_SPLIT_MODALITY.BOXING,
  });

  await handleCreateWorkoutSplit({
    name: 'B',
    workout,
    user,
    modality: WORKOUT_SPLIT_MODALITY.BOXING,
  });

  const context = await getContext({
    graphql: GRAPHQL_TYPE.WEB,
    user,
  });

  const rootValue = {};

  const result = await graphql({
    schema: schemaAdmin,
    source: query,
    rootValue,
    contextValue: context,
  });

  expect(result.errors).toBeUndefined();

  expect(result.data.meWorkoutSplits.edges.length).toEqual(2);
  const [workoutSplit] = result.data.meWorkoutSplits.edges;

  expect(workoutSplit.node.user.id).toEqual(toGlobalId('User', user.id));
  expect(workoutSplit.node.workout.id).toEqual(
    toGlobalId('Workout', workout.id),
  );

  expect(sanitizeTestObject(result.data)).toMatchSnapshot();
});
