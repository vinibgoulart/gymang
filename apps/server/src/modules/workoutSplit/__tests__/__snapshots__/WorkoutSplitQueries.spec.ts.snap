// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get a list of workouts splits 1`] = `
{
  "meWorkoutSplits": {
    "edges": [
      {
        "node": {
          "id": "FROZEN-ID",
          "modality": "BOXING",
          "name": "A",
          "user": {
            "id": "FROZEN-ID",
          },
          "workout": {
            "id": "FROZEN-ID",
            "name": "Workout NaN name",
          },
        },
      },
      {
        "node": {
          "id": "FROZEN-ID",
          "modality": "BOXING",
          "name": "B",
          "user": {
            "id": "FROZEN-ID",
          },
          "workout": {
            "id": "FROZEN-ID",
            "name": "Workout NaN name",
          },
        },
      },
    ],
  },
}
`;
